### PR TITLE
Add useCallback on focus hooks

### LINKED ISSUE #2484: help decrease re-rendering from react-aria hooks?
# 🙋 Feature Request

I'd like to see if react-aria hook could be more friendly and avoid unnecessary re-renders. Specially in hooks that return event handlers. for example `useHover` / `useFocus` / `useFocusRing`

## 🤔 Expected Behavior

returned event handlers such as `onFocus` / `onBlur` and others to be new instances **only when** their dependency changes

## 😯 Current Behavior

as soon as any of the hooks mentioned above are re-called new event handlers are returned.

## 💁 Possible Solution

probably introducing `useCallback` should help. Something like:

```diff
Index: packages/@react-aria/interactions/src/useFocusWithin.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/@react-aria/interactions/src/useFocusWithin.ts b/packages/@react-aria/interactions/src/useFocusWithin.ts
--- a/packages/@react-aria/interactions/src/useFocusWithin.ts	(revision d6b47064ce98bfff50903e6e3147c6f88147de15)
+++ b/packages/@react-aria/interactions/src/useFocusWithin.ts	(date 1634874625411)
@@ -15,7 +15,7 @@
 // NOTICE file in the root directory of this source tree.
 // See https://github.com/facebook/react/tree/cc7c1aece46a6b69b41958d731e0fd27c94bfc6c/packages/react-interactions
 
-import {FocusEvent, HTMLAttributes, useRef} from 'react';
+import {FocusEvent, HTMLAttributes, useCallback, useRef} from 'react';
 
 interface FocusWithinProps {
   /** Whether the focus within events should be disabled. */
@@ -36,45 +36,45 @@
 /**
  * Handles focus events for the target and its descendants.
  */
-export function useFocusWithin(props: FocusWithinProps): FocusWithinResult {
+export function useFocusWithin({onBlurWithin, isDisabled, onFocusWithin, onFocusWithinChange}: FocusWithinProps): FocusWithinResult {
   let state = useRef({
     isFocusWithin: false
-  }).current;
-
-  if (props.isDisabled) {
-    return {focusWithinProps: {}};
-  }
+  });
 
-  let onFocus = (e: FocusEvent) => {
-    if (!state.isFocusWithin) {
-      if (props.onFocusWithin) {
-        props.onFocusWithin(e);
+  let onFocus = useCallback((e: FocusEvent) => {
+    if (!state.current.isFocusWithin) {
+      if (onFocusWithin) {
+        onFocusWithin(e);
       }
 
-      if (props.onFocusWithinChange) {
-        props.onFocusWithinChange(true);
+      if (onFocusWithinChange) {
+        onFocusWithinChange(true);
       }
 
-      state.isFocusWithin = true;
+      state.current.isFocusWithin = true;
     }
-  };
+  }, [onFocusWithin, onFocusWithinChange]);
 
-  let onBlur = (e: FocusEvent) => {
+  let onBlur = useCallback((e: FocusEvent) => {
     // We don't want to trigger onBlurWithin and then immediately onFocusWithin again
     // when moving focus inside the element. Only trigger if the currentTarget doesn't
     // include the relatedTarget (where focus is moving).
-    if (state.isFocusWithin && !e.currentTarget.contains(e.relatedTarget as HTMLElement)) {
-      if (props.onBlurWithin) {
-        props.onBlurWithin(e);
+    if (state.current.isFocusWithin && !e.currentTarget.contains(e.relatedTarget as HTMLElement)) {
+      if (onBlurWithin) {
+        onBlurWithin(e);
+      }
+
+      if (onFocusWithinChange) {
+        onFocusWithinChange(false);
       }
 
-      if (props.onFocusWithinChange) {
-        props.onFocusWithinChange(false);
-      }
+      state.current.isFocusWithin = false;
+    }
+  }, [onBlurWithin, onFocusWithinChange]);
 
-      state.isFocusWithin = false;
-    }
-  };
+  if (isDisabled) {
+    return {focusWithinProps: {}};
+  }
 
   return {
     focusWithinProps: {
```

```diff
Index: packages/@react-aria/focus/src/useFocusRing.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/@react-aria/focus/src/useFocusRing.ts b/packages/@react-aria/focus/src/useFocusRing.ts
--- a/packages/@react-aria/focus/src/useFocusRing.ts	(revision d6b47064ce98bfff50903e6e3147c6f88147de15)
+++ b/packages/@react-aria/focus/src/useFocusRing.ts	(date 1634874799105)
@@ -1,4 +1,4 @@
-import {HTMLAttributes, useState} from 'react';
+import {HTMLAttributes, useCallback, useState} from 'react';
 import {isFocusVisible, useFocus, useFocusVisibleListener, useFocusWithin} from '@react-aria/interactions';
 import {useRef} from 'react';
 
@@ -43,20 +43,20 @@
   let state = useRef({
     isFocused: false,
     isFocusVisible: autoFocus || isFocusVisible()
-  }).current;
+  });
   let [isFocused, setFocused] = useState(false);
-  let [isFocusVisibleState, setFocusVisible] = useState(() => state.isFocused && state.isFocusVisible);
+  let [isFocusVisibleState, setFocusVisible] = useState(() => state.current.isFocused && state.current.isFocusVisible);
 
-  let updateState = () => setFocusVisible(state.isFocused && state.isFocusVisible);
+  let updateState = useCallback(() => setFocusVisible(state.current.isFocused && state.current.isFocusVisible), []);
 
-  let onFocusChange = isFocused => {
-    state.isFocused = isFocused;
+  let onFocusChange = useCallback(isFocused => {
+    state.current.isFocused = isFocused;
     setFocused(isFocused);
     updateState();
-  };
+  }, [updateState]);
 
   useFocusVisibleListener((isFocusVisible) => {
-    state.isFocusVisible = isFocusVisible;
+    state.current.isFocusVisible = isFocusVisible;
     updateState();
   }, [], {isTextInput});
 
@@ -72,7 +72,7 @@
 
   return {
     isFocused,
-    isFocusVisible: state.isFocused && isFocusVisibleState,
+    isFocusVisible: state.current.isFocused && isFocusVisibleState,
     focusProps: within ? focusWithinProps : focusProps
   };
 }
```

## 🔦 Context and 💻 Examples

```
const MyComp = ({ theProp }) => {
  const { isHovered, hoverProps } = useHover({});
  return (
    <>
      <span>{theProp} has nothing to do with hover below</span>
      <h1 {...hoverProps}>Hover me</h1>
      <span>H1 is hovered?: {isHovered}</span>
    </>
  );
};
```
`hoverProps` returns completely new event handlers when `theProp` changes

### PR DESCRIPTION
**Pull request by csantos1113**
Closes #2484 (Partially)

**Partially** because there are many more hooks where we could add `useCallback` and `useMemo` (or fix existing hook dependencies).

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- Filled out test instructions.
- Updated documentation (if it already exists for this component).
- Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

### CODE REVIEW

File: packages/@react-aria/interactions/src/useFocus.ts [hunk 1]
```diff
@@ -33,49 +33,42 @@ interface FocusResult {
...
+      onSyntheticFocus(e);
+    }
+  }, [onFocusChange, onFocusProp, onSyntheticFocus]);
 
   return {
     focusProps: {
-      onFocus,
-      onBlur
+      onFocus: (!isDisabled && (onFocusProp || onFocusChange || onBlurProp)) ? onFocus : undefined,
+      onBlur: (!isDisabled && (onBlurProp || onFocusChange)) ? onBlur : null
```
Comment (snowystinger) [inline packages/@react-aria/interactions/src/useFocus.ts:71]: out of curiosity, why undefined for one but null for the other?

Comment (csantos1113) [reply packages/@react-aria/interactions/src/useFocus.ts:71]: 😬 I kept the logic that was there. not really sure why it was like that

Comment (csantos1113) [reply packages/@react-aria/interactions/src/useFocus.ts:71]: onBlur null: https://github.com/adobe/react-spectrum/pull/3045/files#diff-70aae225f0b9efb36943afd209d8423f070cc51965757699da4ab865b337ee3dL53

onFocus undefined: https://github.com/adobe/react-spectrum/pull/3045/files#diff-70aae225f0b9efb36943afd209d8423f070cc51965757699da4ab865b337ee3dL58

Comment (snowystinger) [reply packages/@react-aria/interactions/src/useFocus.ts:71]: ah, thanks, didn't notice it

**Comment by snowystinger** [comment]
Just a heads up, but lint is failing
```
[1] /home/circleci/react-spectrum/packages/@react-aria/interactions/src/utils.ts
[1]    93:25  error    Unexpected block statement surrounding arrow body; move the returned value immediately after the `=>`                                                                                                                                                            arrow-body-style
[1]    97:18  warning  The ref value 'stateRef.current' will likely have changed by the time this effect cleanup function runs. If this ref points to a node rendered by React, copy 'stateRef.current' to a variable inside the effect, and use that variable in the cleanup function  react-hooks/exhaustive-deps
[1]   118:13  error    'state' is assigned a value but never used   
```

**Comment by csantos1113** [comment]
> Just a heads up, but lint is failing
> 
> ```
> [1] /home/circleci/react-spectrum/packages/@react-aria/interactions/src/utils.ts
> [1]    93:25  error    Unexpected block statement surrounding arrow body; move the returned value immediately after the `=>`                                                                                                                                                            arrow-body-style
> [1]    97:18  warning  The ref value 'stateRef.current' will likely have changed by the time this effect cleanup function runs. If this ref points to a node rendered by React, copy 'stateRef.current' to a variable inside the effect, and use that variable in the cleanup function  react-hooks/exhaustive-deps
> [1]   118:13  error    'state' is assigned a value but never used   
> ```
> 
> you may not need to worry about the warning, but the errors need to be fixed

Fixed, thanks!

**Review by snowystinger — APPROVED** [review]
This is actually a bit easier to read as well. Tested in Chrome, appears to be working. Thanks for having a go at this.

**Comment by LFDanLu** [comment]
GET_BUILD

**Review by LFDanLu — APPROVED** [review]
Did a brief test pass and things look good to me. The team will need to do some in depth testing to make sure there weren't any components that were unintentionally relying on these extra re-renders.

**Comment by csantos1113** [comment]
Mainly kept what was there before 😅

Cesar S
________________________________

**Pull request merged**